Add flexible post processing

andreanne-lemay opened the pull request:
In line with the [post-processing refactoring](https://github.com/ivadomed/ivadomed/issues/464), this PR adds an option for flexible application of post-processing steps. A new parameter is added to the config files: `postprocessing`. The steps enumerated in this parameter will be applied in the given order.

For example, the following will binarize the predictions, remove small objects and threshold with the voxel-wise uncertainty file. 
```json
    "postprocessing": {
        "binarize_prediction": {"thr": 0.5},
        "remove_small": {"unit": "vox", "thr": 3},
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"}
    }
```
Done:
- [x] Add function to parse and apply any postprocessing step
- [x] Modify one config file as example
- [x] Implemented remove small objects
- [x] Implemented binarization
- [x] Implemented uncertainty thresholding
- [x] Noise removal
- [x] Fill holes
- [x] Keep largest objects
- [x] Modify all config files
- [x] Update documentation
- [x] Update tests
- [x] Update scripts
- [x] Add test

andreanne-lemay committed:
Change redundant title

andreanne-lemay committed:
change file name to fit title

andreanne-lemay committed:
models.rst: add sectionning

andreanne-lemay committed:
correct faulty reference

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
Merge branch 'master' of https://github.com/ivadomed/ivadomed

andreanne-lemay committed:
update config.json file for postprocessing

andreanne-lemay committed:
remove testing_params from main.py

andreanne-lemay committed:
remove bin thresholding in testing

andreanne-lemay committed:
rename postprocessing in config file

andreanne-lemay committed:
move remove small objects in postprocessing

andreanne-lemay committed:
method to apply postprocessing steps

andreanne-lemay committed:
change postpro keys to method name

andreanne-lemay committed:
add postprocessing params to evaluate

andreanne-lemay committed:
add uncertainty postprocessing step

andreanne-lemay committed:
chnage remove small object name

andreanne-lemay committed:
add binarization to postprocessing

andreanne-lemay committed:
add binarization to config file

andreanne-lemay committed:
main.py: fix ucnertainty error

andreanne-lemay committed:
fix minor bugs

andreanne-lemay commented:
@charleygros: I started this PR, there is still work to do, but I wanted to make sure we agreed on the format and the way to apply the postprocessing steps. Glad to discuss this and let me know your thoughts!

charleygros reviewed (commented):
Looks good to me.
My only question would be re where `apply_postprocessing` is called.
How do you feel about calling it before `evaluate` --> so that the files that we can visualize (ie in `pred_masks`) are exactly the same as the ones that are evaluated.

andreanne-lemay commented:
> Looks good to me.
> My only question would be re where `apply_postprocessing` is called.
> How do you feel about calling it before `evaluate` --> so that the files that we can visualize (ie in `pred_masks`) are exactly the same as the ones that are evaluated.

The reason why a put the method `apply_postprocessing` in the `__init__` of `Evaluation3DMetrics` is because some postprocessing steps (like remove small objects) require some attributes of this class (e.g., resolution of voxels) and the min_size is needed to compute `target_size`, etc. It was simpler to put it there to access the attributes and modify them during the postprocessing.

andreanne-lemay committed:
add fill_holes postpro step

andreanne-lemay committed:
fill_holes: binarize before

andreanne-lemay committed:
add keep largest objects

andreanne-lemay committed:
add remove noise

andreanne-lemay committed:
put all postprocessing in config.json

andreanne-lemay committed:
apply postpro only if thr >= 0

andreanne-lemay committed:
change all config files

andreanne-lemay committed:
remove testing parameters from tutorials

andreanne-lemay committed:
update config file doc

andreanne-lemay committed:
add test

andreanne-lemay commented:
Sister [PR](https://github.com/ivadomed/data-testing/pull/4) in data-testing

andreanne-lemay committed:
resolve conflicts with master

andreanne-lemay committed:
create postprocessing class

andreanne-lemay committed:
Postprocessing: fix return values

andreanne-lemay committed:
remove postpro from Evaluation3DMetrics

andreanne-lemay committed:
add postprocessing to pred_to_nib

andreanne-lemay committed:
update docstring of utils.py

andreanne-lemay committed:
propagate postprocessing params

andreanne-lemay committed:
modify input required for uncertainty postpro

andreanne-lemay committed:
pred_to_nib: adapt for uncertainty postpro

andreanne-lemay committed:
testing: add one iteration when uncertainty for postpro

andreanne-lemay committed:
remove postprocessing from evaluate

andreanne-lemay committed:
remove postprocessing from evaluation

andreanne-lemay committed:
testing.py: fix minor bugs

andreanne-lemay committed:
fix uncertainty path

andreanne-lemay committed:
update doctstrings

andreanne-lemay committed:
add postprocessing to segment_volume

andreanne-lemay commented:
Changes:
- Create a postprocessing class to manage postprocessing steps (postprocessing.py)
- Before when removing small objects as post processing steps, the small objects where also removed from the GT. I removed this step, since it seems dangerous to modify the GT without assessing the change.
- The postprocessing steps are applied in `pred_to_nib` function, this way the files are saved once. `pred_to_nib` is also used in `segment_volume` (used in SCT), so the postprocessing steps are easily applied when calling this function.
- When uncertainty is applied, I added 1 loop to `run_inference` iteration to save the "_pred.nii.gz" with postprocessing applied. I didn't modify the code in `run_uncertainty`, but I believe we could remove the code that saves the "_pred.nii.gz" file.

andreanne-lemay committed:
auto PEP8

andreanne-lemay committed:
update testing data release version

andreanne-lemay committed:
fix tests

andreanne-lemay committed:
Postprocessing: add condition if no fname

andreanne-lemay committed:
remove dependency to existing models

andreanne-lemay committed:
fix bug on threshold_analysis

charleygros commented on docs/source/configuration_file.rst:
```diff
@@ -404,3 +385,102 @@ Available transformations:
+
+fill\_holes
+^^^^^^^^^^^
+Dict. Fill holes in the predictions. No parameters required (e.i., {}).
```
```suggestion
Dict. Fill holes in the predictions. No parameters required (i.e., {}).
```

charleygros commented on docs/source/configuration_file.rst:
```diff
@@ -404,3 +385,102 @@ Available transformations:
+
+keep\_largest
+^^^^^^^^^^^^^
+Dict. Keeps only the largest object in prediction. Only nearest neighbors are connected to the center,
+diagonally-connected elements are not considered neighbors. No parameters required (e.i, {})
```
```suggestion
Dict. Keeps only the largest connected object in prediction. Only nearest neighbors are connected to the center,
diagonally-connected elements are not considered neighbors. No parameters required (i.e., {})
```

charleygros commented on docs/source/configuration_file.rst:
```diff
@@ -404,3 +385,102 @@ Available transformations:
+
+remove\_noise
+^^^^^^^^^^^^^
+Dict. Sets to zero prediction values under the given threshold ``thr``.
```
```suggestion
Dict. Sets to zero prediction values strictly below the given threshold ``thr``.
```

charleygros commented on docs/source/configuration_file.rst:
```diff
@@ -404,3 +385,102 @@ Available transformations:
+:ref:`uncertainty <Uncertainty>` parameter.
+
+- ``thr``: Float. Threshold is between 0 and 1. Threshold set to ``-1`` will not apply this postprocessing step.
+- ``suffix``: String. Indicates the suffix of an uncertainty file. Choices: ``_unc-vox.nii.gz`` for voxel-wise
+  uncertainty, ``_unc-avgUnc.nii.gz`` for structure-wise uncertainty derived from mean value of ``_unc-vox.nii.gz``
+  within a given connected object, or ``_unc-cv`` for structure-wise uncertainty derived from coefficient of variation.
```
I am guessing you are not mentioning `unc_iou` since the lower the more uncertain. But I am thinking we could still offer the possibility of using it: if iou >> remove vox below thr ; otherwise remove above thr.

andreanne-lemay replied:
I added this possibility in the code + documentation 27bd3cc and e1c5a70

charleygros commented on docs/source/configuration_file.rst:
```diff
@@ -404,3 +385,102 @@ Available transformations:
+targetSize
+^^^^^^^^^^
+- ``unit``: String. Either "vox" for voxels or "mm3". Indicates the unit used to define the target object sizes.
+- ``thr``: List. containing two int values. The two values will create 3 target size bins: minimal size to first list
```
```suggestion
- ``thr``: List. Containing int values. These values will create several consecutive target size bins. For instance with a list of two values, we will have three target size bins: minimal size to first list
```

charleygros replied:
We used it with two values, but in theory the code is code to any values: 4 values --> 5 bins. etc

charleygros commented on docs/source/configuration_file.rst:
```diff
@@ -404,3 +385,102 @@ Available transformations:
+- `config\_spineGeHemis.json <ivadomed/config/config_spineGeHemis.json>`__. Shows how to use the HeMIS-UNet.
+
+- `config\_tumorSeg.json <ivadomed/config/config_tumorSeg.json>`__. Runs a 3D segmentation using a 3D UNet.
+=======
```
Thanks for making the doc much clearer!

charleygros commented on ivadomed/config/config.json:
```diff
@@ -74,19 +74,22 @@
+    "postprocessing": {
+        "remove_noise": {"thr": -1},
+        "keep_largest": {},
+        "binarize_prediction": {"thr": 0.5},
+        "fill_holes": {},
+        "remove_small": {"unit": "vox", "thr": 3},
+        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"}
```
```suggestion
        "binarize_prediction": {"thr": 0.5},
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
        "fill_holes": {},
        "remove_small": {"unit": "vox", "thr": 3}
```

charleygros replied:
I am suggesting this order and the unc thresholding could somewhat induce some holes etc.

charleygros commented on ivadomed/config/config_classification.json:
```diff
@@ -68,19 +68,20 @@
+    },
+    "postprocessing": {
+        "remove_noise": {"thr": -1},
+        "binarize_prediction": {"thr": 0.5},
+        "remove_small": {"unit": "vox", "thr": 3},
+        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"}
```
```suggestion
        "binarize_prediction": {"thr": 0.5},
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
        "remove_small": {"unit": "vox", "thr": 3}
```

charleygros commented on ivadomed/config/config_sctTesting.json:
```diff
@@ -67,19 +67,20 @@
+    "postprocessing": {
+        "remove_noise": {"thr": -1},
+        "binarize_prediction": {"thr": 0.5},
+        "remove_small": {"unit": "vox", "thr": 3},
+        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"}
```
```suggestion
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
        "remove_small": {"unit": "vox", "thr": 3}
```

charleygros commented on ivadomed/config/config_small.json:
```diff
@@ -70,19 +70,20 @@
+    "postprocessing": {
+        "remove_noise": {"thr": -1},
+        "binarize_prediction": {"thr": 0.5},
+        "remove_small": {"unit": "vox", "thr": 3},
+        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"}
```
```suggestion
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
        "remove_small": {"unit": "vox", "thr": 3}
```

charleygros commented on ivadomed/config/config_spineGeHemis.json:
```diff
@@ -83,19 +83,20 @@
+    "postprocessing": {
+        "remove_noise": {"thr": -1},
+        "binarize_prediction": {"thr": 0.5},
+        "remove_small": {"unit": "vox", "thr": 3},
+        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"}
```
```suggestion
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
        "remove_small": {"unit": "vox", "thr": 3}        
```

charleygros commented on ivadomed/config/config_tumorSeg.json:
```diff
@@ -76,19 +76,20 @@
+    "postprocessing": {
+        "remove_noise": {"thr": -1},
+        "binarize_prediction": {"thr": 0.5},
+        "remove_small": {"unit": "vox", "thr": 3},
+        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"}
```
```suggestion
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
        "remove_small": {"unit": "vox", "thr": 3}
```

charleygros commented on ivadomed/config/config_vertebral_labeling.json:
```diff
@@ -69,19 +69,20 @@
+    "postprocessing": {
+        "remove_noise": {"thr": -1},
+        "binarize_prediction": {"thr": 0.5},
+        "remove_small": {"unit": "vox", "thr": 3},
+        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"}
```
```suggestion
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
        "remove_small": {"unit": "vox", "thr": 3}
```

charleygros commented on ivadomed/testing.py:
```diff
@@ -58,21 +59,22 @@ def test(model_params, dataset_test, testing_params, log_directory, device, cuda
     if (testing_params['uncertainty']['epistemic'] or testing_params['uncertainty']['aleatoric']) and \
             testing_params['uncertainty']['n_it'] > 0:
-        n_monteCarlo = testing_params['uncertainty']['n_it']
+        n_monteCarlo = testing_params['uncertainty']['n_it'] + 1
```
eheh :+1:

charleygros commented on ivadomed/testing.py:
```diff
@@ -58,21 +59,22 @@ def test(model_params, dataset_test, testing_params, log_directory, device, cuda
-    # COMPUTE UNCERTAINTY MAPS
-    if n_monteCarlo > 1:
-        imed_utils.run_uncertainty(ifolder=path_3Dpred)
+        # If uncertainty computation, don't apply it on last iteration for prediction
+        if testing_params['uncertainty']['applied'] and (n_monteCarlo - 2 == i_monteCarlo):
+            testing_params['uncertainty']['applied'] = False
+            # COMPUTE UNCERTAINTY MAPS
+            imed_utils.run_uncertainty(ifolder=path_3Dpred)
```
I am not sure to understand the reason for the extra MC iteration. Could you please help me?

andreanne-lemay replied:
With the old code, if there were only 2 MC iterations, `run_inference` would only be run twice and the pred files would be generated in `run_uncertainty`. To avoid code duplication related to applying the postprocessing steps in `run_uncertainty`, an extra iteration is added to go through `run_inference` a last time to generate `_pred.nii.gz` files with postprocessing. With this current code, I think we could remove the code related to the generation of the `_pred.nii.gz` in `run_uncertainty`.

charleygros commented on testing/test_script.py:
```diff
@@ -141,13 +142,19 @@ def test_create_eval_json():
+            "remove_noise": {"thr": 0.01},
+            "keep_largest": {},
+            "binarize_prediction": {"thr": 0.5},
+            "fill_holes": {},
+            "remove_small": {"unit": "vox", "thr": 3},
+            "uncertainty": {"thr": 0.4, "suffix": "_unc-vox.nii.gz"}
```
```suggestion
            "uncertainty": {"thr": 0.4, "suffix": "_unc-vox.nii.gz"},
            "fill_holes": {},
            "remove_small": {"unit": "vox", "thr": 3}
```

charleygros reviewed (commented):
Reviewed all files except `evaluation.py` and `postprocessing.py`. Will do a bit later.

andreanne-lemay committed:
Update docs/source/configuration_file.rst

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
Update docs/source/configuration_file.rst

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
Update docs/source/configuration_file.rst

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
Update docs/source/configuration_file.rst

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
Update ivadomed/config/config.json

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
Update ivadomed/config/config_vertebral_labeling.json

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
Update ivadomed/config/config_tumorSeg.json

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
Update testing/test_script.py

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
add uncertainty thresholding on _unc-iov and _soft

andreanne-lemay committed:
update ucnertainty pp documentation with iou and soft

charleygros commented on ivadomed/postprocessing.py:
```diff
@@ -292,7 +292,10 @@ def uncertainty(self, thr, suffix):
             if os.path.exists(uncertainty_path):
                 data_uncertainty = nib.load(uncertainty_path).get_fdata()
-                self.data_pred = mask_predictions(self.data_pred, data_uncertainty < thr)
+                if suffix == "_unc-iou.nii.gz" or suffix == "_soft.nii.gz":
+                    self.data_pred = mask_predictions(self.data_pred, data_uncertainty > thr)
```
I was not thinking of using `_soft.nii.gz`.. but it could make sense! And that's clearly documented so.. all good :)

charleygros commented on ivadomed/postprocessing.py:
```diff
@@ -207,3 +212,133 @@ def label_file_from_coordinates(nifti_image, coord_list):
+       """
+        if thr >= 0:
+            mask = self.data_pred > thr
+            self.data_pred = mask_predictions(self.data_pred, mask)
```
Looks great!! :100:

andreanne-lemay committed:
Merge remote-tracking branch 'origin' into al/post-processing
